File: RxTools-library/src/main/java/com/vondear/rxtools/view/progress/sprite/ShapeSprite.java
package com.vondear.rxtools.view.progress.sprite;

import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.ColorFilter;
import android.graphics.Paint;

/**
 * Created by ybq.
 */
public abstract class ShapeSprite extends Sprite {

    private Paint mPaint;
    private int mUseColor;
    private int mBaseColor;

    public ShapeSprite() {
        setColor(Color.WHITE);
        mPaint = new Paint();
        mPaint.setAntiAlias(true);
        mPaint.setColor(mUseColor);
    }

    @Override
    public void setColor(int color) {
        mBaseColor = color;
        updateUseColor();
    }

    @Override
    public int getColor() {
        return mBaseColor;
    }

    @SuppressWarnings("unused")
    public int getUseColor() {
        return mUseColor;
    }

    @Override
    public void setAlpha(int alpha) {
        super.setAlpha(alpha);
        updateUseColor();
    }

    private void updateUseColor() {
        int alpha = getAlpha();
        alpha += alpha >> 7;
        final int baseAlpha = mBaseColor >>> 24;
        final int useAlpha = baseAlpha * alpha >> 8;
        mUseColor = (mBaseColor << 8 >>> 8) | (useAlpha << 24);
    }

    @Override
    public void setColorFilter(ColorFilter colorFilter) {
        mPaint.setColorFilter(colorFilter);
    }

    @Override
    protected final void drawSelf(Canvas canvas) {
        mPaint.setColor(mUseColor);
        drawShape(canvas, mPaint);
    }

    public abstract void drawShape(Canvas canvas, Paint paint);
}

File: RxUI/src/main/java/com/tamsiree/rxui/view/progressing/sprite/ShapeSprite.kt
package com.tamsiree.rxui.view.progressing.sprite

import android.graphics.Canvas
import android.graphics.Color
import android.graphics.ColorFilter
import android.graphics.Paint

/**
 * @author tamsiree
 */
abstract class ShapeSprite : Sprite() {
    private val mPaint: Paint
    var useColor = 0
        private set
    private var mBaseColor = 0
    override var color: Int
        get() = mBaseColor
        set(color) {
            mBaseColor = color
            updateUseColor()
        }

    override fun setAlpha(alpha: Int) {
        super.setAlpha(alpha)
        updateUseColor()
    }

    private fun updateUseColor() {
        var alpha = alpha
        alpha += alpha shr 7
        val baseAlpha = mBaseColor ushr 24
        val useAlpha = baseAlpha * alpha shr 8
        useColor = mBaseColor shl 8 ushr 8 or (useAlpha shl 24)
    }

    override fun setColorFilter(colorFilter: ColorFilter?) {
        mPaint.colorFilter = colorFilter
    }

    override fun drawSelf(canvas: Canvas?) {
        mPaint.color = useColor
        drawShape(canvas, mPaint)
    }

    abstract fun drawShape(canvas: Canvas?, paint: Paint?)

    init {
        color = Color.WHITE
        mPaint = Paint()
        mPaint.isAntiAlias = true
        mPaint.color = useColor
    }
}